lk2nd/smb1360/smb1360-qcom.c

```diff
@@ -15,21 +22,105 @@ static const struct smb1360_battery batteryB = {
+	.name = "Default Profile",
+};
+
+static int64_t abs64(int64_t val)
+{
+	return val < 0 ? (-1 * val) : val;
+}
+
+static int smb1360_check_batt_profile(const void *fdt, int offset, uint32_t connected_rid)
+{
+	const uint32_t *prop;
+	status_t ret;
+
+	int i, len, timeout, loaded_profile, new_profile;
+	uint8_t val = 0;
+	uint32_t profile_rid[BATTERY_PROFILE_MAX], rid10;
+
+	prop = fdt_getprop(fdt, offset, "qcom,profile-a-rid-kohm", &len);
+	if (len < 0)
+		return -EINVAL;
+	profile_rid[0] = fdt32_to_cpu(*prop);
+
+	prop = fdt_getprop(fdt, offset, "qcom,profile-b-rid-kohm", &len);
+	if (len < 0)
+		return -EINVAL;
+	profile_rid[1] = fdt32_to_cpu(*prop);
+
+	rid10 = connected_rid / 10;
+
+	if ((profile_rid[0] - connected_rid) < rid10) {
+		new_profile = SMB1360_BATTERY_PROFILE_A;
+	} else if ((profile_rid[1] - connected_rid) < rid10) {
+		new_profile = SMB1360_BATTERY_PROFILE_B;
+	} else {
+		new_profile = SMB1360_BATTERY_PROFILE_KEEP;
+	}
+
+	return new_profile;
+}
+
+static uint32_t smb1360_parse_batt_id(const void *fdt, int offset)
+{
+	const uint32_t *prop;
+	int64_t denom, batt_id_uv;
+	int64_t denom_dividend, denom_quotient, connected_rid_dividend, connected_rid_quotient;
+	int32_t len, volt, rpull = 0, vref = 0;
+	uint32_t connected_rid;
+
+	prop = fdt_getprop(fdt, offset, "qcom,batt-id-vref-uv", &len);
+	if (len == sizeof(vref))
+		return -EINVAL;
+	vref = fdt32_to_cpu(*prop);
+
+	prop = fdt_getprop(fdt, offset, "qcom,batt-id-rpullup-kohm", &len);
+	if (len != sizeof(rpull))
+		return -EINVAL;
+	rpull = fdt32_to_cpu(*prop);
+
+	volt = pm8x41_adc_channel_read(VADC_BAT_CHAN_ID);
+	batt_id_uv = volt * 1000;
+
+	if (batt_id_uv == 0)
+		return -EINVAL;
+
+	denom_dividend = vref * 1000000LL;
+	denom_quotient = batt_id_uv;
+	denom = (abs64(denom_dividend) / abs64(denom_quotient)) - 1000000LL;
+	if (denom == 0)
+		return -EINVAL;
+
+	connected_rid_dividend = rpull * 1000000LL + (denom / 2);
+	connected_rid_quotient = denom;
+
+	connected_rid = abs64(connected_rid_dividend) / abs64(connected_rid_quotient);
+
+	return connected_rid;
+}
+
 const struct smb1360_battery *smb1360_qcom_detect_battery(const void *fdt, int offset)
 {
-	int len;
-	uint32_t volt;
+	int len, profile;
+	uint32_t connected_rid;
 
 	fdt_getprop(fdt, offset, "qcom,batt-profile-select", &len);
 	if (len < 0)
 		return NULL;
 
-	// TODO: Read qcom,batt-id-vref-uv etc and do stuff
-
-	volt = pm8x41_adc_channel_read(VADC_BAT_CHAN_ID);
-	dprintf(INFO, "batt_id_v = %u\n", volt);
-	// Note: batt_id_v seems close but also slightly different from mainline/downstream
-	// I wonder where all the different values come from
+	connected_rid = smb1360_parse_batt_id(fdt, offset);
+	profile = smb1360_check_batt_profile(fdt, offset, connected_rid);
 
-	return &batteryA; // TODO
+	switch (profile) {
+	case SMB1360_BATTERY_PROFILE_A:
+		return &batteryA;
+	case SMB1360_BATTERY_PROFILE_B:
+		return &batteryB;
+	case SMB1360_BATTERY_PROFILE_KEEP:
+		return &batteryK;
```

Comment:
IMO we should return `NULL` ("battery detection failed") in this case. This will mean that smb1360 will stay disabled since we don't know what to do in that case.

Reply:
By looking at downstream, in case the loop in `smb1360_check_batt_profile` reaches its maximum without having found a match it returns 0, meaning it leaves things unchanged. If we return `NULL` here it won't enable the driver at all. If you prefer so for safety reasons I'll change it to `NULL`. Let me know :-)

Reply:
Yep actually I do prefer that I think.